activesupport/lib/active_support/message_verifier.rb

```diff
@@ -64,54 +66,184 @@ def valid_message?(signed_message)
+          raise InvalidSignature
+        end
+      end
     end
 
-    # Generates a signed message for the provided value.
+    # Generates a signed message for the given +value+ and +options+.
     #
     # The message is signed with the +MessageVerifier+'s secret. Without knowing
-    # the secret, the original value cannot be extracted from the message.
+    # the secret, the original value cannot be extracted from the message. A
+    # message is signed as a JSON Web Token (JWT). The final message is
+    # concatenation of base64-encoded JWS header, base64-encoded JWT claims set
+    # and base64-encoded signature separated by a period ('.'). This format is
+    # known as a JSON Web Signature (JWS):
+    # <base64-encoded header>.<base64-encoded claims>.<base64-encoded signature>
+    #
+    # JSON Web Token (JWT) - A string representing a set of claims as a JSON
+    # object that is encoded in a JWS or JWE, enabling the claims to be
+    # signed and/or encrypted.
+    #
+    # JWT Header - A JSON object that describes the cryptographic operations
+    # applied to the JWT. When the JWT is signed, the JWT Header is a
+    # JWS Header. A JWS header signed using 'SHA1' algorithm:
+    #
+    #   { 'typ' => 'JWT', 'alg' => 'SHA1' }
+    #
+    # JWT claims set - A JSON object that contains the claims conveyed by
+    # the JWT. For our purpose, we limit our claims set to consist of
+    # +pld+ (payload), +exp+ (expiry), and +for+ (purpose/scope). This becomes
+    # the payload of JWS. For example:
+    #
+    #   { 'pld' => 'a private message', 'for' => 'test', 'exp' => '2015-06-29T07:16:12.579Z' }
+    #
+    # JWS Signature - Signature over the JWS Header and the JWS Payload (Claims set).
+    # The signing input is given by ASCII(BASE64URL(UTF8(JWS Protected Header)) || '.' ||
+    # BASE64URL(JWS Payload)).
+    #
+    # ==== Options
+    #
+    # * <tt>:expires_at</tt> - Explicit expiring time for the signed message.
+    # * <tt>:expires_in</tt> - Relative time the message should
+    #   expire after (e.g. 1.hour).
+    # * <tt>:for</tt> - The signed message purpose confines usage to places with
+    #   the same purpose. Defaults to a universally usable message.
+    #
+    # ==== Examples
+    #
+    # Messages can be signed to ensure that the data hasn't been tampered with.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
-    #   verifier.generate 'a private message' # => "BAhJIhRwcml2YXRlLW1lc3NhZ2UGOgZFVA==--e2d724331ebdee96a10fb99b089508d1c72bd772"
-    def generate(value)
-      data = encode(@serializer.dump(value))
-      "#{data}--#{generate_digest(data)}"
+    #   verifier.generate('a private message')
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIOdW5pdmVyc2FsBjsGVDoIZXhwSSIdMjAxNS0wOC0wOVQxODowNjowMy44NzhaBjsGVA==.ZmRlYjhlMTkxNGI0NGQ4OWRkNDdjZWM5YTY1NmQ3MTM5OWY5NzRmMQ=="
```

Comment:
Yeah, it would be better to go with ellipsis here and some other places too. I will make this change :+1: .